---
title: Bam processing
author: Robert Policastro
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Bam processing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Create tsrexplorer object.

```{r message=FALSE}
library("TSRexploreR")
# BAM file.
bam_file <- system.file("extdata", "S288C.bam", package="TSRexploreR")

# Genome assembly.
assembly <- system.file("extdata", "S288C_Assembly.fasta", package="TSRexploreR")

# Sample sheet.
samples <- data.frame(sample_name="S288C", file_1=bam_file, file_2=NA)

exp <- tsr_explorer(sample_sheet=samples, genome_assembly=assembly)
```

Import BAMs.

```{r message=FALSE, warning=FALSE}
exp <- import_bams(exp, paired=TRUE, proper_pair=TRUE)
```

Plot soft-clipped histogram.

```{r message=FALSE, fig.height=2, fig.width=3}
library(ggplot2)
softclip_histogram(exp) +
  theme_bw() +
  scale_fill_viridis_d()
```

Plot softclipped base frequency.

```{r fig.height=2.5, fig.width=3}
softclip_composition(exp) +
  theme_bw() +
  scale_fill_viridis_d()
```

Correct for G content.

```{r message=FALSE}
exp <- G_correction(exp)
```

Aggregate TSSs by position.

```{r message=FALSE}
exp <- tss_aggregate(exp)
```
